internal/services/controllers/workflows/queue.go

```diff
@@ -603,79 +603,107 @@ func (wc *WorkflowsControllerImpl) getLock(key string) *sync.Mutex {
 
-	runningWorkflowRuns, err := wc.repo.WorkflowRun().ListWorkflowRuns(ctx, tenantId, &repository.ListWorkflowRunsOpts{
-		WorkflowVersionId: &workflowVersionId,
-		GroupKey:          &groupKey,
-		Statuses:          &[]db.WorkflowRunStatus{running},
-		OrderBy:           repository.StringPtr("createdAt"),
-		OrderDirection:    repository.StringPtr("ASC"),
-	})
 	if err != nil {
-		return fmt.Errorf("could not list running workflow runs: %w", err)
-	}
-
-	queuedWorkflowRuns, err := wc.repo.WorkflowRun().ListWorkflowRuns(ctx, tenantId, &repository.ListWorkflowRunsOpts{
-		WorkflowVersionId: &workflowVersionId,
-		GroupKey:          &groupKey,
-		Statuses:          &[]db.WorkflowRunStatus{queued},
-		OrderBy:           repository.StringPtr("createdAt"),
-		OrderDirection:    repository.StringPtr("ASC"),
-	})
-	if err != nil {
-		return fmt.Errorf("could not list queued workflow runs: %w", err)
+		return fmt.Errorf("could not pop workflow runs: %w", err)
 	}
 
-	runningCount := len(runningWorkflowRuns.Rows)
-	queuedCount := len(queuedWorkflowRuns.Rows)
-
-	// Calculate how many runs we need to cancel
-	toCancel := max(0, runningCount+queuedCount-maxRuns)
-
 	// Cancel the oldest running workflows
-	for i := 0; i < toCancel && i < runningCount; i++ {
-		row := runningWorkflowRuns.Rows[i]
-		workflowRunId := sqlchelpers.UUIDToStr(row.WorkflowRun.ID)
+	for i := range toCancel {
+		row := toCancel[i]
+		workflowRunId := sqlchelpers.UUIDToStr(row.ID)
 
 		err = wc.cancelWorkflowRun(ctx, tenantId, workflowRunId)
 
 		if err != nil {
 			return fmt.Errorf("could not cancel workflow run: %w", err)
 		}
+	}
+
+	for i := range toStart {
+		row := toStart[i]
+		workflowRunId := sqlchelpers.UUIDToStr(row.ID)
+		queuedStepRuns, err := wc.repo.WorkflowRun().QueueWorkflowRunJobs(ctx, tenantId, workflowRunId)
+
+		if err != nil {
+			return fmt.Errorf("could not queue workflow run jobs: %w", err)
+		}
+
+		g := new(errgroup.Group)
+		for _, stepRun := range queuedStepRuns {
+			stepRunCp := stepRun
+
+			g.Go(func() error {
+				return wc.mq.AddMessage(
+					ctx,
+					msgqueue.JOB_PROCESSING_QUEUE,
+					tasktypes.StepRunQueuedToTask(stepRunCp),
+				)
+			})
+		}
 
-		err := wc.cancelWorkflowRunJobs(ctx, tenantId, workflowRunId, "CANCEL_IN_PROGRESS")
+		err = g.Wait()
 
 		if err != nil {
-			return fmt.Errorf("queueByCancelInProgress: could not cancel workflow run jobs: %w", err)
+			return fmt.Errorf("could not start workflow run: %w", err)
 		}
+	}
+
+	return nil
+}
+
+func (wc *WorkflowsControllerImpl) queueByCancelNewest(ctx context.Context, tenantId string, workflowVersion *dbsqlc.GetWorkflowVersionForEngineRow) error {
+	ctx, span := telemetry.NewSpan(ctx, "queue-by-cancel-newest")
+	defer span.End()
+
+	workflowId := sqlchelpers.UUIDToStr(workflowVersion.WorkflowVersion.WorkflowId)
+	maxRuns := int(workflowVersion.ConcurrencyMaxRuns.Int32)
+
+	mutex := wc.getLock(fmt.Sprintf("%s:%s", tenantId, workflowId))
```

Comment:
ditto with this mutex, are we protected by the locks in the DB?

Reply:
Removed